### PR TITLE
Compile the minipal as an object and static library

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Compile the repo minipal as an object and static library for usage across the product as a single unit instead of separate source files.

### CODE REVIEW

File: src/native/libs/CMakeLists.txt
Comment (am11) [inline src/native/libs/CMakeLists.txt:5]: Maybe change the name to LibsNative (i.e. `-s libs.native`) while at it.

File: src/native/libs/CMakeLists.txt
Comment (am11) [inline src/native/libs/CMakeLists.txt:7]: Why is this condition needed now and wasn't needed before? Is it fixing some pre-existing warning?

File: src/native/minipal/CMakeLists.txt
Comment (am11) [inline src/native/minipal/CMakeLists.txt:19]: nit: EOF new line is missing

File: src/coreclr/nativeaot/Runtime/Full/CMakeLists.txt
Comment (SingleAccretion) [inline src/coreclr/nativeaot/Runtime/Full/CMakeLists.txt:31]: Nit: add it to `Runtime/Portable/CMakeLists.txt` too.

File: src/coreclr/nativeaot/Runtime/Full/CMakeLists.txt
Comment (SingleAccretion) [inline src/coreclr/nativeaot/Runtime/Full/CMakeLists.txt:65]: Does CMake flow `CLR_CONTROL_FLOW_GUARD ON` to the minipal sources automagically?

File: src/native/libs/CMakeLists.txt
Comment (jkoritzinsky) [reply src/native/libs/CMakeLists.txt:7]: We used to have an include guard in `configurecompiler.cmake` as it could get included twice from here when the libs CMake is included in the CoreCLR CMake tree. Adding this condition allows us to remove the include guard.

File: src/coreclr/nativeaot/Runtime/Full/CMakeLists.txt
Comment (jkoritzinsky) [reply src/coreclr/nativeaot/Runtime/Full/CMakeLists.txt:65]: Good catch, I'll set that up.

File: src/native/minipal/random.c
Comment (vcsjones) [inline src/native/minipal/random.c:100]: Are these return codes correct? Looking at other returns, it seems like `0` is success and `-1` is failure.

File: src/mono/mono/mini/CMakeLists.txt
Comment (am11) [inline src/mono/mono/mini/CMakeLists.txt:340]: Might look better on line 338: `$<TARGET_OBJECTS:minipal_objects>`

**Comment by jkoritzinsky** [comment]
This PR is ready for review

**Review by am11 — APPROVED** [review]
@jkoritzinsky, this is looking great! 👍 

Just curious, did you see any difference in libcoreclr.so/dylib size before/after (release configuration)? I guess since we don't compile with --icf=all/safe, with objects LTO should improve the size a bit.

**Comment by MichalStrehovsky** [comment]
I triggered a rt-sz run on this at https://github.com/MichalStrehovsky/rt-sz/actions/runs/11230071292/job/31217408645 and it's failing with errors like:

```
  /home/runner/work/rt-sz/rt-sz/microsoft.netcore.app.runtime.linux-x64/Release/runtimes/linux-x64/native/libSystem.Native.a(pal_random.c.o): in function `SystemNative_GetNonCryptographicallySecureRandomBytes':
  pal_random.c:(.text.SystemNative_GetNonCryptographicallySecureRandomBytes+0x6): undefined reference to `minipal_get_non_cryptographically_secure_random_bytes'
```

I don't really know much about what this actually changes in the outputs (don't know what's an object and static library to begin with), but there might be trouble with the outputs when using different linkers or different versions of linkers (this repo uses a pretty recent LLD but all customer scenarios are going to default to some old binutils that comes with the distro, rt-sz included).

Random other rt-sz run from yesterday completed without issues: https://github.com/MichalStrehovsky/rt-sz/issues/77

I'm not saying we should block this on rt-sz, but... don't leave on vacation right after merging this.

**Comment by jkotas** [comment]
> I don't really know much about what this actually changes in the outputs (don't know what's an object and static library to begin with),

The minipal is linked in `Runtime.WorkstationGC/ServerGC.a`. I think that the problem is that the native libraries shims that depend on minipal as well are specified after `Runtime.WorkstationGC/ServerGC.a`. This ordering does not work for some linkers. Unix linkers expect that the unresolved symbol in one library is going to satisfied by a library that follows. Unix linkers do not always go back to try to satisfy the symbols.

https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets#L154-L155 was a different instance of the same ordering issue.

**Comment by am11** [comment]
@MichalStrehovsky, could you shed some light on why the smoke tests didn’t fail? Are they using the LKG package instead of the just-built nupkg? It might be beneficial to set up a similar GitHub Actions workflow in the dotnet/runtime repo for NativeAOT changes, like the one in the rt-sz repo, to quickly assess live builds. Given that millions of workflows run on GitHub Actions daily, I believe it can handle the scale, especially since it failed in just 13 minutes, which is a relatively quick turnaround! 🙂

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky, could you shed some light on why the smoke tests didn’t fail? Are they using the LKG package instead of the just-built nupkg?

I don't have much else than what I already wrote above - "this repo uses a pretty recent LLD but all customer scenarios are going to default to some old binutils that comes with the distro, rt-sz included".

The action simply runs dotnet publish on whatever is `ubuntu-latest` on GitHub, replacing ILC/framework/AotSdk (corelib+libs.native+runtime) with whatever was just built. This results in using binutils to link stuff. Runtime repo and smoke tests in it use LLD - we had to switch to that due to supply chain hardening. We touched on this test hole at https://github.com/dotnet/runtime/issues/98688#issuecomment-1963769495.

> Given that millions of workflows run on GitHub Actions daily, I believe it can handle the scale, especially since it failed in just 13 minutes, which is a relatively quick turnaround!

It does that in 13 minutes by only building a tiny portion of the repo. It doesn't even bother building packs. It is a fragile configuration that often fails when things shuffle. The moment one tries to do it "properly", it stops being cheap (both time-wise, and maintenance cost (compliance) wise). Runtime repo is not very suitable for end to end testing.

**Comment by am11** [comment]
> `ubuntu-latest` on GitHub

Yes, `ubuntu-latest` currently has llvm14 (they are rolling out llvm18 in next two to three weeks). I've tested and the error caused by this PR persists with llvm18 as well. I normally use the [prereq docker](https://github.com/am11/CrossRepoCITesting/blob/256d46b9a4a48ef220f8cc9778431a7b57ad48af/.github/workflows/build-arm.yml#L31) (same as runtime CI) to not rely on GH host.

> The moment one tries to do it "properly", it stops being cheap

`./build.sh clr+libs+packs -c Release` takes half hour in GitHub actions which is less that the time to complete an average windows leg. I'm using the action with `sudo rm -rf "$AGENT_TOOLSDIRECTORY";` to make room for tri-subset build.

**Comment by am11** [comment]
> `ubuntu-latest` on GitHub

Yes, `ubuntu-latest` currently has llvm14 (they are rolling out llvm18 in next two to three weeks). I've tested and the error caused by this PR persists with llvm18 as well. I normally use the [prereq docker](https://github.com/am11/CrossRepoCITesting/blob/256d46b9a4a48ef220f8cc9778431a7b57ad48af/.github/workflows/build-arm.yml#L31) (same as runtime CI) to not rely on GA host.

> The moment one tries to do it "properly", it stops being cheap

`./build.sh clr+libs+packs -c Release` takes half hour in GitHub actions which is less that the time to complete an average windows leg. I'm using the action with `sudo rm -rf "$AGENT_TOOLSDIRECTORY";` to make room for tri-subset build.

**Comment by MichalStrehovsky** [comment]
> Yes, `ubuntu-latest` currently has llvm14 (they are rolling out llvm18 in next two to three weeks).

Is LLVM version relevant for this conversation? The action leaves everything at defaults and we default to bfd (the actions log conveniently has `/usr/bin/ld.bfd: /usr/bin/ld.bfd: DWARF error: invalid or unhandled FORM value: 0x23`, confirming it's bfd).

**Comment by am11** [comment]
> Is LLVM version relevant for this conversation?

I agree that version is not relevant here. With LinkerFlavor=bfd it fails with LinkerFalvor=lld it works. I thought we have a project or two using bfd in the repo but apparently not.

**Comment by am11** [comment]
(overall it's a size improvement https://github.com/am11/rt-sz/issues/2#issuecomment-2401410742 👍)

**Comment by jkoritzinsky** [comment]
I can try linking the minipal into libSystem.Native as well, but I think I'll get duplicate symbol errors. I'll throw it at CI and see what it says.

**Comment by jkotas** [comment]
> I can try linking the minipal into libSystem.Native as well, but I think I'll get duplicate symbol errors. 

I think you want to tweak the order of libraries passed to the linker and/or ship minipal as a separate library.

**Comment by jkoritzinsky** [comment]
Yeah I'll probably end up having to ship it as a separate library. I was hoping not to, but that's likely the easiest option all-around.

**Comment by am11** [comment]
> I'll get duplicate symbol errors.

rt-sz and /ba are green. We only get duplicate errors without `PRIVATE` modifier, which we have, so it's all good. 👍

**Comment by jkotas** [comment]
/ba is not green. Mono AOT tests are failing with:

```
/tmp/helix/working/A4E308C2/p/build/common/LibraryBuilder.targets(28,5): error MSB4018: duplicate symbol '_minipal_get_length_utf8_to_utf16' in: [/private/tmp/helix/working/A4E308C2/w/B6300A50/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/A4E308C2/p/build/common/LibraryBuilder.targets(28,5): error MSB4018:     /tmp/helix/working/A4E308C2/w/B6300A50/e/publish/libmonosgen-2.0.a(utf8.c.o) [/private/tmp/helix/working/A4E308C2/w/B6300A50/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/A4E308C2/p/build/common/LibraryBuilder.targets(28,5): error MSB4018:     /tmp/helix/working/A4E308C2/w/B6300A50/e/publish/libSystem.Native.a(utf8.c.o) [/private/tmp/helix/working/A4E308C2/w/B6300A50/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/A4E308C2/p/build/common/LibraryBuilder.targets(28,5): error MSB4018: ld: 14 duplicate symbols for architecture arm64 [/private/tmp/helix/working/A4E308C2/w/B6300A50/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/A4E308C2/p/build/common/LibraryBuilder.targets(28,5): error MSB4018: clang: error: linker command failed with exit code 1 (use -v to see invocation) 
```

File: src/mono/mono/eglib/CMakeLists.txt
Comment (am11) [reply src/mono/mono/eglib/CMakeLists.txt:65]: eglib doesn't need to expose minipal symbols through it. eglib_api exposes stuff for mono public API (src/native/public) for embedded, so it's different.

Comment (jkoritzinsky) [reply src/mono/mono/eglib/CMakeLists.txt:65]: This dependency is public as eglib_objects does depend on the minipal, so other places in Mono that use eglib_objects needs the minipal. That's why it's `PUBLIC`.

Comment (am11) [reply src/mono/mono/eglib/CMakeLists.txt:65]: > other places in Mono that use eglib_objects needs the minipal

Those other places being:

```
src/mono/mono/component/CMakeLists.txt
src/mono/mono/mini/CMakeLists.txt
src/mono/mono/profiler/CMakeLists.txt
```

can we not link minipal there as we have done in multiple coreclr projects?

Comment (jkoritzinsky) [reply src/mono/mono/eglib/CMakeLists.txt:65]: If we add it there, that's the exact same as adding it here, just more verbose and further from usage. We don't get any improvements by using `PRIVATE` in 4 places instead of `PUBLIC` in one.

I'm trying to move to using `PUBLIC` in more places where it's correct to reduce duplication and move our CMake to be more target-oriented instead of basically just throwing code at the wall and seeing what works. Mono's still quite a mess with the `TARGET_OBJECTS` usage, but at some point I'll clean that up.

Comment (am11) [reply src/mono/mono/eglib/CMakeLists.txt:65]: Overall, this PR is size improvement over main branch on NativeAOT side. Which means whatever was happening before with duplicate C sources in multiple projects, the linker is doing better job deduplicating the static libs in final binaries. Exposing those symbols via some other object would not impact the size. Mono is generally heavy on static libraries.

**Comment by am11** [comment]
> `Reorder libraries on the command line when linking for Unix.`

Now that we switched to ILC's LKG packages, will it help? Fixing mono is pretty simple in comparison. :)

File: src/mono/mono/eglib/CMakeLists.txt
Comment (jkoritzinsky) [reply src/mono/mono/eglib/CMakeLists.txt:65]: `PUBLIC` doesn't expose symbols via different objects, it just specifies how dependencies are added to CMake targets. It's not a size concern in the final product, but a concern in the cleanliness and understandability of the infrastructure.

**Comment by jkoritzinsky** [comment]
Given how much I've had to fight mono to get symbols working right here in the first place, I really don't think fixing mono is easier. In fact, I see it as significantly more difficult to fix that various different build configurations (it's really easy to end up with a configuration in Mono where you don't have the symbols available at all while also having one where you have symbol duplicates).

File: src/mono/mono/eglib/CMakeLists.txt
Comment (am11) [reply src/mono/mono/eglib/CMakeLists.txt:65]: If changing PUBLIC to PRIVATE and adding minipal to the three cmakelists files does not fix the issue, I understand that this is not the solution.

But if it does (due to the linkage details), what is the problem? I think it is rather 'better' this way because:
1. this is exactly what's been done in coreclr
2. is it an anti-pattern of sorts what we have in coreclr? I think this is debatable since we were thinking about cleaning up eglib during the utf8 work and calling minipal_utf8.. APIs directly where needed (mono/mini etc.). From that topology, it's currently a local concern of eglib that it make use of some minipal APIs. It is not exactly a 'complete PAL' for mono akin coreclr-pal.

Comment (jkoritzinsky) [reply src/mono/mono/eglib/CMakeLists.txt:65]: What I'm saying is I don't think changing `PUBLIC` to `PRIVATE` here would actually solve the problem.

The problem is that the `minipal_objects` library is linked into both the Mono runtime (which needs it to be in the static lib for when it statically links everything together for a native app) and into System.Native (which needs it as it depends on minipal as well). I don't want to try to thread the minipal static lib everywhere it needs to ship (see the brotli PR of mine that I still haven't finished due to this exact issue). In general, I'm trying to defer to linking the minipal objects into the runtime static libs and having the other static libs (that don't directly depend on the runtime) use the symbols from there.

This problem is unrelated to the `minipal` target mentioned here. The one mentioned here ensures that any Mono shared libs that depend on eglib_objects have the `minipal` static lib also linked into them as that's a transitive dependency.

**Comment by jkoritzinsky** [comment]
I think I got the library order right now. Can I get another round of reviews?

File: src/native/minipal/CMakeLists.txt
Comment (AaronRobinsonMSFT) [inline src/native/minipal/CMakeLists.txt:19]: Can we comment why we need this build?

File: src/native/minipal/CMakeLists.txt
Comment (AaronRobinsonMSFT) [inline src/native/minipal/CMakeLists.txt:16]: Why do we need to specify this argument? I thought we were consuming the default configurations from `eng/`?

File: src/native/minipal/sansupport.c
Comment (AaronRobinsonMSFT) [inline src/native/minipal/sansupport.c:8]: Do we need these declarations and definitions?

File: src/native/minipal/CMakeLists.txt
Comment (jkoritzinsky) [reply src/native/minipal/CMakeLists.txt:16]: I was thinking about how the nativeaot's tree changes the default (NativeAOT defaults to `OFF`). We don't need to be explicit here because this is always included from the CMake root (where we have the right defaults)

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**
